### DISCUSSION TITLE
BIND view implementation suggestion

### DISCUSSION THREAD

**Discussion by kemeris2000** [sys4/netbox-plugin-dns#13] [Ideas]
I case you use BIND views to implement split-horizon for external/internal networks, netbox-plugin-dns is suitable as you usually need only two zones. However netbox-plugin-dns is not practical when you need to implement DNS load balancing by client's country, as you  need one view with domain zone for each country in the worlds (~195).
So I propose to allow multiple views assignment to DNS zone record itself, rather than to DNS zone only and generate all needed DNS zones for selected views.

For example, let's say I have statically defined zone domain1.com assigned to Default view in netbox-plugin-dns.

<pre>
views:
  - Default
  - CO.UK
  - DE

zones:
  - <b>view: Default
    type: static</b>
    name: domain1.com
    nameservers:
      - ns1.domain1.com
      - ns2.domain1.com
    managed_records:
      - '86400	SOA	ns1.domain1.com. admin.domain1.com. 1683128316 172800 7200 259200'
      - '86400	NS	ns1.domain1.com.'
      - '86400	NS	ns2.domain1.com.'
    records:
       - name: www
         type: A
         <b>value: 5.5.5.5</b>
         <b>views: []</b>
       - name: www
         type: A
         <b>value: 12.12.12.12</b>
         <b>views:
           - CO.UK
           - DE</b>
</pre>

netbox-plugin-dns could generate the following result. This would simplify definition of BIND "match-clients" directive for each view:

<pre>

zones:
  - <b>view: Default
    type: static</b>
    name: domain1.com
    nameservers:
      - ns1.domain1.com
      - ns2.domain1.com
    managed_records:
      - '86400	SOA	ns1.domain1.com. admin.domain1.com. 1683128316 172800 7200 259200'
      - '86400	NS	ns1.domain1.com.'
      - '86400	NS	ns2.domain1.com.'
    records:
      - name: www
         type: A
         <b>value: 5.5.5.5</b>
  - <b>view: CO.UK
    type: dynamic</b>
    name: domain1.com
    nameservers:
      - ns1.domain1.com
      - ns2.domain1.com
    managed_records:
      - '86400	SOA	ns1.domain1.com. admin.domain1.com. 1683128316 172800 7200 259200'
      - '86400	NS	ns1.domain1.com.'
      - '86400	NS	ns2.domain1.com.'
    records:
      - name: www
         type: A
         <b>value: 12.12.12.12</b>
  - <b>view: DE
    type: dynamic</b>
    name: domain1.com
    nameservers:
      - ns1.domain1.com
      - ns2.domain1.com
    managed_records:
      - '86400	SOA	ns1.domain1.com. admin.domain1.com. 1683128316 172800 7200 259200'
      - '86400	NS	ns1.domain1.com.'
      - '86400	NS	ns2.domain1.com.'
    records:
      - name: www
         type: A
         <b>value: 12.12.12.12</b>
</pre>

Let me know what you think?

**Comment by peteeckel** [comment]
Hi @kemeris2000, that's really an interesting idea, thanks for this input!

Currently your goal would be difficult to achieve, as views are implemented as a kind of 'namespace' concept that completely separates zones from each other. That is, however, the opposite of what your use case is, so the concept doesn't fit. While that's the case pretty often, usually that isn't a problem because in most cases split-horizon is, as you noted, used for a handful of views where the duplication doesn't matter much.

If I see it correctly, in your case a large part of the zone is identical for all views, and just some records (e.g. `www`) are specific to a view. That would mean duplicating large numbers of zones and records for just very few records that are specific to a view, which is clearly not optimal.

**Question to the community: Is this something that is of wider interest so it justifies the (probably non-trivial) effort of implementing it in the plugin itself?**

Putting a zone into multiple views would be one possible solution. The 'views as namespaces' concept would need to be dropped and replaced by some inheritance concept (allowing zones to inherit records from their counterparts in 'parent' views and adding records or overwriting inherited records with their own). I think that would be the cleanest solution. The hardest part will be the correct generation of `PTR` records, especially if the view/zone structure is not the same for the reverse zones.

~~A simpler solution, though certainly not as elegant, would be the introduction of the standard NetBox 'Config Context' field to NetBox DNS objects. That would make it possible to add, in your example, the country specific records in a suitable data structure attached to a zone, and then handle the country specific stuff during zone file generation, e.g. in your export script. While admittedly lacking some elegance this could be implemented rather quickly and without changing much, and it could also be used for other purposes. Would that be useful for you, maybe as a short-term solution?~~

~~Thinking of this, I'll probably implement it anyway - it won't hurt and probably add some value for others as well.~~

_Edit: It seems that the 'Config Context' functionality is at least not supported for Plugins (yet). So it's not as short-term as I was hoping ..._

**Comment by kemeris2000** [comment]
I have started to implement this feature on netbox custom script level.
Thank you for suggestions [peteeckel](https://github.com/peteeckel).
